Tests/Common/Orders/Fills/EquityFillModelTests.cs

```diff
@@ -516,54 +518,234 @@ public void PerformsStopMarketFillSell()
+            {
+                new Tick(time, Symbols.SPY, saleCondition, "P", 100, 1m),
+                new Tick(time, Symbols.SPY, 1m, 0.9m, 1.1m)
+            }, typeof(Tick));
+
+            fill = model.MarketOnOpenFill(equity, order);
+            Assert.AreEqual(0, fill.FillQuantity);
+
+            const decimal expected = 1m;
             // market opens after 30min
             time = reference.AddMinutes(30);
             TimeKeeper.SetUtcDateTime(time.ConvertToUtc(TimeZones.NewYork));
-            security.SetMarketPrice(new TradeBar(time, Symbols.SPY, 1.45m, 2.0m, 1.1m, 1.40m, 100));
 
-            fill = model.MarketOnOpenFill(security, order);
+            // The quote is received after the market is open, but the trade is not
+            equity.Update(new List<Tick>
+            {
+                new Tick(time.AddMinutes(-1), Symbols.SPY, saleCondition, "P", 100, expected),
+                new Tick(time, Symbols.SPY, 1m, 0.9m, 1.1m)
+            }, typeof(Tick));
+
+            fill = model.MarketOnOpenFill(equity, order);
+            Assert.AreEqual(0, fill.FillQuantity);
+
+            // One quote and some trades with different conditions are received after the market is open,
+            // but there is trade prior to that with different price
+            equity.Update(new List<Tick>
+            {
+                new Tick(time.AddMinutes(-1), Symbols.SPY,  "80000001", "P", 100, 0.9m),   // Not Open
+                new Tick(time, Symbols.SPY, saleCondition, "Q", 100, 0.95m),            // Open but not primary exchange
+                new Tick(time, Symbols.SPY, saleCondition, "P", 100, expected),         // Fill with this tick
+                new Tick(time, Symbols.SPY, 1m, 0.9m, 1.1m),
+                new Tick(time, Symbols.SPY,  "80000001", "P", 100, 0.95m),   // Open but not primary exchange
+            }, typeof(Tick));
+
+            fill = model.MarketOnOpenFill(equity, order);
+            Assert.AreEqual(order.Quantity, fill.FillQuantity);
+            Assert.AreEqual(expected, fill.FillPrice);
+        }
+
+        [TestCase(Resolution.Minute, 3, 17, 0, 9, 29)]
+        [TestCase(Resolution.Minute, 4, 0, 0, 9, 29)]
+        [TestCase(Resolution.Minute, 4, 8, 0, 9, 29)]
+        [TestCase(Resolution.Minute, 4, 9, 30, 9, 29, true)]
+        [TestCase(Resolution.Hour, 3, 17, 0, 8, 29)]
+        [TestCase(Resolution.Hour, 4, 0, 0, 8, 0)]
+        [TestCase(Resolution.Hour, 4, 8, 0, 8, 0)]
+        [TestCase(Resolution.Hour, 4, 9, 30, 8, 0, true)]
+        [TestCase(Resolution.Daily, 3, 17, 0, 8, 0)]
+        [TestCase(Resolution.Daily, 4, 0, 0, 8, 0)]
+        [TestCase(Resolution.Daily, 4, 8, 0, 8, 0)]
+        [TestCase(Resolution.Daily, 4, 9, 30, 8, 0)]
+        public void PerformsMarketOnOpenUsingOpenPriceWithDifferentOrderSubmissionDateTime(Resolution resolution, int day, int hour, int minute, int ref_hour, int ref_minute, bool nextDay = false)
+        {
+            var model = new EquityFillModel();
```

Comment:
Nice!
Would be nice to add these tests I think, where ref time is after market open too:
```
        [TestCase(Resolution.Minute, 4, 9, 30, 9, 31, true)]
        [TestCase(Resolution.Hour, 4, 9, 30, 11, 0, true)]
```


> var model = new EquityFillModel();

Don't think should be required? Can do `equity.FillModel`? Closer to the real use case too 👍